apiserver/facades/client/modelupgrader/controllerupgrader.go

```diff
@@ -0,0 +1,154 @@
+package modelupgrader
+
+import (
+	"context"
+	"fmt"
+	"github.com/juju/juju/apiserver/common"
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/core/agentbinary"
+	coreerrors "github.com/juju/juju/core/errors"
+	corelogger "github.com/juju/juju/core/logger"
+	"github.com/juju/juju/core/semversion"
+	controllerupgradererrors "github.com/juju/juju/domain/controllerupgrader/errors"
+	"github.com/juju/juju/domain/modelagent"
+	modelagenterrors "github.com/juju/juju/domain/modelagent/errors"
+	internalerrors "github.com/juju/juju/internal/errors"
+	"github.com/juju/names/v6"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/rpc/params"
+)
+
+// ControllerUpgraderService mirrors the controller upgrader service
+// so we can easily mock it for unit tests.
+type ControllerUpgraderService interface {
+	UpgradeController(ctx context.Context) (semversion.Number, error)
+	UpgradeControllerWithStream(
+		ctx context.Context,
+		stream modelagent.AgentStream,
+	) (semversion.Number, error)
+	UpgradeControllerToVersion(
+		ctx context.Context,
+		desiredVersion semversion.Number,
+	) error
+	UpgradeControllerToVersionAndStream(
+		ctx context.Context,
+		desiredVersion semversion.Number,
+		stream modelagent.AgentStream,
+	) error
+}
+
+// ControllerUpgraderAPI upgrades a controller and a model hosting the controller.
+type ControllerUpgraderAPI struct {
+	authorizer facade.Authorizer
+	check      common.BlockCheckerInterface
+
+	upgraderService ControllerUpgraderService
+
+	logger corelogger.Logger
+}
+
+// NewControllerUpgraderAPI instantiates a new [ControllerUpgraderAPI].
+func NewControllerUpgraderAPI(
+	authorizer facade.Authorizer,
+	check common.BlockCheckerInterface,
+	upgraderService ControllerUpgraderService,
+	logger corelogger.Logger,
+) *ControllerUpgraderAPI {
+	return &ControllerUpgraderAPI{
+		authorizer:      authorizer,
+		check:           check,
+		upgraderService: upgraderService,
+		logger:          logger,
+	}
+}
+
+// AbortModelUpgrade returns not supported, as it's not possible to move
+// back to a prior version.
+func (c *ControllerUpgraderAPI) AbortModelUpgrade(ctx context.Context, arg params.ModelParam) error {
+	return errors.NotSupportedf("abort model upgrade")
```

Comment:
Can we use the new errors pacakge for this. The import path is `github.com/juju/juju/internal/errors`.

Something like this be great: `return errors.New("aborting model upgrades is not supported").Add(coreerrors.NotSupported)`